Modules/Filtering/FFT/include/itkComplexToComplex1DFFTImageFilter.hxx

```diff
@@ -0,0 +1,170 @@
+/*=========================================================================
+ *
+ *  Copyright NumFOCUS
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef itkComplexToComplex1DFFTImageFilter_hxx
+#define itkComplexToComplex1DFFTImageFilter_hxx
+
+#include "itkComplexToComplex1DFFTImageFilter.h"
+
+#include "itkVnlComplexToComplex1DFFTImageFilter.h"
+
+#if defined(ITK_USE_FFTWD) || defined(ITK_USE_FFTWF)
+#  include "itkFFTWComplexToComplex1DFFTImageFilter.h"
+#endif
+
+#if defined(ITKUltrasound_USE_clFFT)
+#  include "itkOpenCLComplexToComplex1DFFTImageFilter.h"
+#endif
+
+#include "itkMetaDataDictionary.h"
+#include "itkMetaDataObject.h"
+
+namespace itk
+{
+
+template <typename TInputImage, typename TOutputImage>
+typename ComplexToComplex1DFFTImageFilter<TInputImage, TOutputImage>::Pointer
+ComplexToComplex1DFFTImageFilter<TInputImage, TOutputImage>::New()
+{
+  Pointer smartPtr = ObjectFactory<Self>::Create();
+
+#ifdef ITKUltrasound_USE_clFFT
+  if (smartPtr.IsNull())
+  {
+    smartPtr =
+      dynamic_cast<Self *>(OpenCLComplexToComplex1DFFTImageFilter<TInputImage, TOutputImage>::New().GetPointer());
+  }
```

Comment:
Thanks for doing this.

Please see the discussion in Pull Request #2744, which is about a change that is being deferred to a later version of ITK but it highlights an issue that is applicable here: in current ITK code generally, a non-null `MyClass::Pointer` created from `ObjectFactory<MyClass>::Create()` will require a call to `UnRegister()` before being returned, but such a smart pointer created from `MyClass::New().GetPointer()` will require that `UnRegister()` not be called.

Reply:
Thank you @Leengit ! I will give this issue attention before marking the PR as ready for review.

Reply:
@Leengit is it sufficient to call `UnRegister` immediately after `ObjectFactory<Self>::Create()` returns? i.e. [`itkComplexToComplex1DFFTImageFilter.hxx`](https://github.com/InsightSoftwareConsortium/ITK/pull/2766/files#diff-3d993d90d493d78a45746642e1032067aff037768f0a833c855c25d960699c8bR39-R45):

```  
Pointer smartPtr = ObjectFactory<Self>::Create();

if (smartPtr.IsNotNull())
{
  // Decrement ITK SmartPointer produced from object factory
  smartPtr->UnRegister();
}
```

Reply:
#2744 is a pending PR, which might not be merged soon. Write the code so it works now. That PR will need some update, and expansion now that these classes are being ported into ITK proper. On the positive side, there will be less changes in remote modules.

Reply:
Thanks @dzenanz, I meant to ask if the code as currently written with a call to `UnRegister` is compatible with the current state of the ITK code base not taking #2744 into account.

The plan is to eventually move to using the object factory to generate FFT classes based on library availability (FFTW, OpenCL) rather than the dispatch methods here, which are particularly affected by #2744.